src/dryrun.ts

```diff
@@ -0,0 +1,128 @@
+ * from (transaction.SignedTransaction) and creates a DryrunRequest object with relevant balances
+ * @param client - the AlgodClient to make requests against
+ * @param txns - the array of SignedTransaction to use for generating the DryrunRequest object
+ * @param protocolVersion - the string representing the protocol version to use
+ * @param latestTimestamp - the timestamp
+ * @returns the DryrunRequest object constructed from the SignedTransactions passed
+ */
+export async function createDryrun(
+  client: AlgodClient,
+  txns: SignedTransaction[],
+  protocolVersion?: string,
+  latestTimestamp?: number | bigint,
+  round?: number | bigint,
+  sources?: DryrunSource[]
+): Promise<DryrunRequest> {
+  const appInfos = [];
+  const acctInfos = [];
+
+  const apps = [];
+  const assets = [];
+  const accts = [];
+  for (const t of txns) {
+    if (t.txn.type === TransactionType.appl) {
+      accts.push(t.txn.from);
+
+      if (t.txn.appAccounts) accts.push(...t.txn.appAccounts);
+
+      if (t.txn.appForeignApps) apps.push(...t.txn.appForeignApps);
+
+      if (t.txn.appForeignAssets) assets.push(...t.txn.appForeignAssets);
+
+      // Create application,
+      if (t.txn.appIndex === 0) {
+        const appId = 1;
+        appInfos.push(
+          new Application(
+            appId,
+            new ApplicationParams({
+              creator: encodeAddress(t.txn.from.publicKey),
+              approvalProgram: t.txn.appApprovalProgram,
+              clearStateProgram: t.txn.appClearProgram,
+              localStateSchema: new ApplicationStateSchema(
+                t.txn.appLocalInts,
+                t.txn.appLocalByteSlices
+              ),
+              globalStateSchema: new ApplicationStateSchema(
+                t.txn.appGlobalInts,
+                t.txn.appGlobalByteSlices
+              ),
+            })
+          )
+        );
+      } else {
+        apps.push(t.txn.appIndex);
+      }
+    }
+  }
+
+  // Dedupe and add creator to accts array
+  const assetPromises = [];
+  for (const assetId of [...new Set(assets)]) {
+    assetPromises.push(
+      client
+        .getAssetByID(assetId)
+        .do()
+        .then((assetInfo) => {
+          accts.push(assetInfo.params.creator);
+        })
+    );
+  }
+  await Promise.all(assetPromises);
+
+  // Dedupe and get app info for all apps
+  const appPromises = [];
+  for (const appId of [...new Set(apps)]) {
+    client
+      .getApplicationByID(appId)
+      .do()
+      .then((appInfo) => {
+        const ai = { ...appInfo };
+        ai.params['approval-program'] = Buffer.from(
+          appInfo.params['approval-program'],
+          'base64'
+        );
+        ai.params['clear-state-program'] = Buffer.from(
+          appInfo.params['clear-state-program'],
+          'base64'
+        );
+        appInfos.push(ai);
+      });
+  }
+  await Promise.all(appPromises);
```

Comment:
`appPromises` never gets anything added to it. Same thing for `acctPromises` below.

Note since these 3 stages are all independent, it would be better to just `await Promise.all([... assetPromises, ...appPromises, ...acctPromises])` at the end, instead of an await after each of the 3.